### PR TITLE
Added isAssoc to test if array is associative

### PR DESCRIPTION
**Pull request by m1**

### CODE REVIEW

**Comment by bocharsky-bw** [comment]
@m1 I'm just wondering for what purpose do you need this `isAssoc()` method and where do you want to use it? I'm interesting in real use cases of it.

File: src/AbstractArray.php [hunk 1]
```diff
@@ -357,6 +357,22 @@ public function isEmpty()
     }
 
     /**
+     * Check whether array is associative or not
+     *
+     * @return bool Returns true if associative, false otherwise
+     */
+    public function isAssoc()
+    {
+        foreach (array_keys($this->elements) as $key) {
```
Comment (bocharsky-bw) [inline src/AbstractArray.php]: Use `$this->getKeys()` instead. We already have method for it.

**Comment by m1** [comment]
@bocharsky-bw Sometimes APIs (for example Amazon Product Advertising API) return indexed arrays if there are more than one result or associative arrays if the result is more than one e.g.

```
"ItemLookupResponse": {
        "Item": {
            "Asin": "B00008OE6I",
            "ItemAttributes": {
                "Example": "Example"
            }
        }
    }
```

OR

```
"ItemLookupResponse": {
        "Item": [
            {
                "Asin": "B00008OE6I",
                "ItemAttributes": {
                    "Example": "Example"
                }
            },
            {
                "Asin": "B00008OE6I",
                "ItemAttributes": {
                    "Example": "Example"
                }
            }
        ]
    }
```

File: src/AbstractArray.php [hunk 2]
```diff
@@ -357,6 +357,22 @@ public function isEmpty()
...
+     */
+    public function isAssoc()
+    {
+        foreach ($this->getKeys() as $key) {
+            if (!is_int($key)) {
+                return true;
+            }
+        }
+
+        return false;
```
Comment (bocharsky-bw) [inline src/AbstractArray.php]: @m1 My only point is that I don't consider next array as associative:

``` php
[
    1 => 'one',
    'two' => 2,
    3 => 'three',
]
```

It's a mixed (numeric and assoc).

So if we want to check whether array is only **assoc**, lets inverse this method logic to:

``` php
foreach ($this->getKeys() as $key) {
    if (!is_string($key)) {
        return false;
    }
}

return true;
```

What do you think?

BTW, we could add `isNumeric()` method to check whether array is only **numeric** with:

``` php
foreach ($this->getKeys() as $key) {
    if (!is_int($key)) {
        return false;
    }
}

return true;
```

Is it sounds good for you?

**Comment by m1** [comment]
@bocharsky-bw should there be a isMixed() or do you think isAssoc() and isNumeric() are sufficient?

**Comment by bocharsky-bw** [comment]
I think it'll be overhead. Lets stop on this two methods now.

File: src/AbstractArray.php [hunk 3]
```diff
@@ -357,6 +357,46 @@ public function isEmpty()
...
 
     /**
+     * Check whether array is associative or not
+     *
+     * @return bool Returns true if associative, false otherwise
+     */
+    public function isAssoc()
+    {
+        if ($this->isEmpty()) {
+            return false;
```
Comment (bocharsky-bw) [inline src/AbstractArray.php:367]: :+1: for this check, I missing it...

**Comment by bocharsky-bw** [comment]
@m1 I like it! Could you squash all your comments to a single one?

**Comment by m1** [comment]
@bocharsky-bw What comments?

**Comment by bocharsky-bw** [comment]
Oh, sorry, I meant _commits_.

**Comment by m1** [comment]
Okay, do you want me to open another pull req with it?

**Comment by bocharsky-bw** [comment]
No, I just want to squash your commits here. You could squash them and then do:
`git push origin +master`
to force pushing squashed commits.

**Comment by bocharsky-bw** [comment]
Don't worry about it much, if you hard to do it, I could do it myself.

**Comment by m1** [comment]
@bocharsky-bw Do you want to do it? Also how would you feel about a case-sensitive/insensitive switch for `contains()` and `containsKey()`, e.g. `contains($string [,$case_insensitive = false])`

```
$a = A::create(['a', 'b', 'c']);
$a->contains('C') // false
$a->contains('C', false); // true
```

**Comment by bocharsky-bw** [comment]
OK, no problem, I'll do it!

**Comment by bocharsky-bw** [comment]
I closed this PR after merge it and rebase. But don't worry because I've reused your commit so you'll get the full credit for your work.Thanks!

**Comment by bocharsky-bw** [comment]
What about `case_insensitive`, I don't like an idea to overload this methods with new parameters and complex a method logic. I'll plan to add `exists()` method with a closure as a first argument, so we could to use any custom logic we need. It's similar to Doctrine's [ArrayCollection::exists](https://github.com/doctrine/collections/blob/v1.3.0/lib/Doctrine/Common/Collections/ArrayCollection.php#L195)

**Comment by m1** [comment]
For what it's worth case insensitive search on an array isn't that complex, would be like one or two lines at max. But `exists()` sounds good nethertheless!

**Comment by bocharsky-bw** [comment]
I vote to add `exists()` method and you could create a PR for it if you want.

**Pull request closed without merging**